common-utils/src/common/deployment/__init__.py

```diff
@@ -356,125 +378,270 @@ class FlowSpec(BaseModel):
+                            "awslogs-stream-prefix": task_name,
+                        },
+                    },
+                }
+            ],
+            "requiresCompatibilities": [
+                "FARGATE",
+            ],
+            "cpu": default_cpu,
+            "memory": default_memory,
+            "ephemeralStorage": {"sizeInGiB": ephemeral_storage},
+        }
+        # check to see if we have any registered task definitions
+        task_def_arns = ecs_client.list_task_definitions(familyPrefix=task_name, maxResults=5)  # type: ignore
+        # if we have some then we treat this as an existing task definition and start diff
+        if task_def_arns["taskDefinitionArns"]:
+            current_task_def_response = ecs_client.describe_task_definition(  # type: ignore
+                taskDefinition=task_name
+            )
+            current_task_def = current_task_def_response["taskDefinition"]
+            final_task_def_arn = current_task_def["taskDefinitionArn"]
+            # if any of our top level keys have changed we mark as "register = True"
+            if not all(
+                new_task_def_dict.get(key) == current_task_def.get(key)
+                for key in task_def_compare_keys
+            ):
+                LOGGER.info(
+                    f"Task definition state has changed for {task_name}, registering new definition..."
+                )
+                register = True
+            # if any of our container definition keys have changed we mark as "register = True"
+            elif not all(
+                new_task_def_dict.get("containerDefinitions", [{}])[0].get(key)
+                == current_task_def.get("containerDefinitions", [{}])[0].get(key)
+                for key in container_def_compare_keys
+            ):
+                LOGGER.info(
+                    f"Task definition state has changed for {task_name}, registering new definition..."
+                )
+                register = True
+        # if no registered arns, we treat this a brand new and mark  "register = True" for initial registration
+        else:
+            LOGGER.info(
+                f"Task definition not registered for {task_name}, registering new definition..."
             )
+            register = True
+        # if "register = True" we register a task definition
+        if register:
+            response = ecs_client.register_task_definition(**new_task_def_dict)  # type: ignore
+            final_task_def_arn = response["taskDefinition"]["taskDefinitionArn"]
         else:
-            # return image string
-            return create_image_name_str(ENVIRONMENT_TYPE, PYPROJECT_METADATA.docker_envs[v]["python_version"])  # type: ignore
+            LOGGER.info(
+                f"Task definition state has not changed {task_name}, new revision not needed..."
+            )
+        # this will return the current or newly registered task definition version arn
+        return final_task_def_arn
```

Comment:
This is a brand new method.  The prefect-aws ECSTask class makes a lot of assumptions about what the registered task definition should look like that I do not 100% agree with.

ultimately I want a very vanilla and specific task definition with deployment overrides passed the boto3 task_run method.  This is why we are using a custom version of this library for now.

This method will analyze specific keys and determine if a new task definition needs to be registered.

Reply:
This method also takes in the account_id and ecs_client, which is inline with only calling out to AWS as needed.